src/core/server/http/lifecycle/auth.ts

```diff
@@ -26,79 +27,101 @@ enum ResultType {
-  public isAuthenticated() {
-    return this.type === ResultType.authenticated;
-  }
-  public isRedirected() {
-    return this.type === ResultType.redirected;
-  }
-  public isRejected() {
-    return this.type === ResultType.rejected;
-  }
+interface Authenticated {
+  type: ResultType.authenticated;
+  state: object;
 }
 
+interface Redirected {
+  type: ResultType.redirected;
+  url: string;
+}
+
+interface Rejected {
+  type: ResultType.rejected;
+  error: Error;
+  statusCode?: number;
+}
+
+type AuthResult = Authenticated | Rejected | Redirected;
+
+const authResult = {
+  authenticated(state: object): AuthResult {
+    return { type: ResultType.authenticated, state };
+  },
+  redirected(url: string): AuthResult {
+    return { type: ResultType.redirected, url };
+  },
+  rejected(error: Error, options: { statusCode?: number } = {}): AuthResult {
+    return { type: ResultType.rejected, error, statusCode: options.statusCode };
+  },
+  isValid(candidate: any): candidate is AuthResult {
+    return (
+      candidate &&
+      (candidate.type === ResultType.authenticated ||
+        candidate.type === ResultType.rejected ||
+        candidate.type === ResultType.redirected)
+    );
+  },
+  isAuthenticated(result: AuthResult): result is Authenticated {
+    return result.type === ResultType.authenticated;
+  },
+  isRedirected(result: AuthResult): result is Redirected {
+    return result.type === ResultType.redirected;
+  },
+  isRejected(result: AuthResult): result is Rejected {
+    return result.type === ResultType.rejected;
+  },
+};
+
 /**
  * @public
  * A tool set defining an outcome of Auth interceptor for incoming request.
  */
 export interface AuthToolkit {
   /** Authentication is successful with given credentials, allow request to pass through */
-  authenticated: (credentials: any) => AuthResult;
+  authenticated: (state: object) => AuthResult;
   /** Authentication requires to interrupt request handling and redirect to a configured url */
   redirected: (url: string) => AuthResult;
   /** Authentication is unsuccessful, fail the request with specified error. */
   rejected: (error: Error, options?: { statusCode?: number }) => AuthResult;
 }
 
 const toolkit: AuthToolkit = {
-  authenticated: AuthResult.authenticated,
-  redirected: AuthResult.redirected,
-  rejected: AuthResult.rejected,
+  authenticated: authResult.authenticated,
+  redirected: authResult.redirected,
+  rejected: authResult.rejected,
 };
 
 /** @public */
 export type AuthenticationHandler<T> = (
-  request: Request,
+  request: Readonly<Request>,
   sessionStorage: SessionStorage<T>,
   t: AuthToolkit
-) => Promise<AuthResult>;
+) => AuthResult | Promise<AuthResult>;
 
 /** @public */
 export function adoptToHapiAuthFormat<T = any>(
   fn: AuthenticationHandler<T>,
-  sessionStorage: SessionStorageFactory<T>
+  sessionStorage: SessionStorageFactory<T>,
+  onSuccess: (req: Request, state: unknown) => void = noop
```

Comment:
Any practical reason to widen the type of `state` from object to `unknown` here? I'd expect them to match.

Reply:
they aren't. state contain authorization information which shouldn't be cached in cookies.
that was discussed with @legrego  in DM
> Right now, the cookie only stores authentication information, and we delegate all authorization to Elasticsearch at “runtime”. Storing this info in a cookie would mean caching authorization details, which we don’t want to do.